python/vowpalwabbit/pyvw.py

```diff
@@ -1157,6 +1159,59 @@ def __str__(self):
             ["{}:{}:{}".format(c.action, c.cost, c.probability) for c in self.costs]
         )
 
+class ccb_label_type(enum.Enum):
+    UNSET = pylibvw.vw.tUNSET
+    SHARED = pylibvw.vw.tSHARED
+    ACTION = pylibvw.vw.tACTION
+    SLOT = pylibvw.vw.tSLOT
+
+class slot_outcome:
+  cost: float = 0
+  action_probs: List[Tuple[int, float]]
+  
+  def __init__(self, cost, action_probs, ex=None):
+    if isinstance(ex, example):
+        self.from_example(ex)
+    else:
+        self.cost = cost
+        self.action_probs = action_probs
+
+  def from_example(self, ex):
+    self.cost = ex.get_ccb_cost()
+    self.action_probs = []
+    for i in range(ex.get_ccb_num_included_actions()):
+        self.action_probs.append((ex.get_ccb_class(i), ex.get_ccb_probs(i)))
+
+  def __str__(self):
+    top_action, top_prob = self.action_probs[0]
+    out = f"{top_action}:{top_prob}:{self.cost}"
+    for action, prob in self.action_probs[1:]:
+        out += f",{action}:{prob}"
+    return out
+
+class ccb_label(abstract_label):
+  type: ccb_label_type = ccb_label_type.UNSET
+  explicit_included_actions: List[int] = []
+  weight: float = 0
+  outcome: Optional[slot_outcome] = None
+"""Class for conditional contextual bandits VW label"""
+
+    def __init__(self, type: ccb_label_type, explicit_included_actions: List[int], weight: float, outcome: Optional[slot_outcome], ex=None):
+        abstract_label.__init__(self)
+        if isinstance(ex, example):
```

Comment:
@jackgerrits @bassmang  I'm a little confused here, how do I keep an optional `example` as an argument here. This way user can extract information directly from example. I don't think python supports multiple constructors. Let me know your thoughts!

Reply:
In the same way other label types do it. For example here: https://github.com/VowpalWabbit/vowpal_wabbit/blob/6318d17f36599cab70902edf7de1dc142538c326/python/vowpalwabbit/pyvw.py#L1038

`label` is the first arg. It is either a `float` or an `example`. This just makes use of the fact it is a positional argument.

Reply:
I was thinking of  this approach:

```python
def __init__(self, type, ...., **kwargs):
  ex = kwargs.get('ex')
  if isinstance(ex, example):
      ....
```

Reply:
Correct me if I'm wrong, you're asking whether we should allow single positional arg input to labels (and expecting that to be `example` class`). 

I think nope, this method is much more clear in its intent.